### PR TITLE
Add getters for raw private/public keys

### LINKED ISSUE #6259: It's not possible to export the raw public key for X25519/Ed255919/X448/Ed448
Other algorithms provide low level APIs for getting hold of low level details such as raw key formats. However X5519/Ed25519/X448/Ed448 do not provide this capability (although you can *import* the raw key formats). Arguably this is a bug that should have been done at the same time as the import code.

This causes real problems in practice, e.g. see:

https://stackoverflow.com/questions/50351287/how-to-openssl-1-1-1-ecdh-with-25519/

### PR DESCRIPTION
**Pull request by mattcaswell**
When we added support for setting raw private/public keys we should have added getters at the same time. This fixes that omission.

Fixes #6259.

### CODE REVIEW

File: crypto/asn1/ameth_lib.c [hunk 1]
```diff
@@ -417,3 +417,19 @@ void EVP_PKEY_asn1_set_set_pub_key(EVP_PKEY_ASN1_METHOD *ameth,
 {
     ameth->set_pub_key = set_pub_key;
 }
+
+void EVP_PKEY_asn1_set_get_priv_key(EVP_PKEY_ASN1_METHOD *ameth,
```
Comment (richsalz) [inline crypto/asn1/ameth_lib.c]: Why is "asn1" in the name?

File: doc/man3/EVP_PKEY_new.pod [hunk 2]
```diff
@@ -63,13 +71,31 @@ creation of a CMAC in the B<cipher> argument.
 EVP_PKEY_new_mac_key() works in the same way as EVP_PKEY_new_raw_private_key().
 New applications should use EVP_PKEY_new_raw_private_key() instead.
 
+EVP_PKEY_get_raw_private_key() fills the buffer provided by B<priv> with raw
+private key data. The number of bytes written is populated in B<*len>. If the
+buffer B<priv> is NULL then B<*len> is populated with the number of bytes
+required in the buffer. The calling application is responsible for ensuring that
```
Comment (richsalz) [inline doc/man3/EVP_PKEY_new.pod]: s/in the buffer/hold the key/ ?

File: doc/man3/EVP_PKEY_new.pod [hunk 3]
```diff
@@ -63,13 +71,31 @@ creation of a CMAC in the B<cipher> argument.
...
+the buffer is large enough to receive the private key data. This function only
+works for algorithms that support raw private keys. Currently this is:
+B<EVP_PKEY_HMAC>, B<EVP_PKEY_POLY1305>, B<EVP_PKEY_SIPHASH>, B<EVP_PKEY_X25519>,
+B<EVP_PKEY_ED25519>, B<EVP_PKEY_X448> or B<EVP_PKEY_ED448>.
+
+EVP_PKEY_get_raw_public_key() fills the buffer provided by B<pub> with raw
+public key data. The number of bytes written is populated in B<*len>. If the
+buffer B<pub> is NULL then B<*len> is populated with the number of bytes
+required in the buffer. The calling application is responsible for ensuring that
```
Comment (richsalz) [inline doc/man3/EVP_PKEY_new.pod]: s/in the buffer/hold the key/ ?

File: doc/man3/EVP_PKEY_new.pod [hunk 4]
```diff
@@ -63,13 +71,31 @@ creation of a CMAC in the B<cipher> argument.
...
+
 =head1 NOTES
 
 The B<EVP_PKEY> structure is used by various OpenSSL functions which require a
 general private key without reference to any particular algorithm.
 
-The structure returned by EVP_PKEY_new() is empty. To add a private key to this
-empty structure the functions described in L<EVP_PKEY_set1_RSA(3)> should be
+The structure returned by EVP_PKEY_new() is empty. To add a private or public
+key to this empty structure the functions described in L<EVP_PKEY_set1_RSA(3)>,
```
Comment (richsalz) [inline doc/man3/EVP_PKEY_new.pod]: active voice: "to add ... keys to this structure, use the appropriate functions from ..."

File: test/evp_extra_test.c [hunk 5]
```diff
@@ -476,6 +477,97 @@ static int test_EVP_PKCS82PKEY(void)
...
+struct keys_st {
+    int type;
+    char *priv;
+    char *pub;
+} keys[] = {
+    {
+        EVP_PKEY_HMAC,
+        "0123456789",
+        NULL
+    }, {
```
Comment (richsalz) [inline test/evp_extra_test.c]: This indentation is kinda wonky.  I think most entries fit in one line, except those with big keys.

File: test/evp_extra_test.c [hunk 6]
```diff
@@ -581,6 +673,7 @@ int setup_tests(void)
 #ifndef OPENSSL_NO_EC
     ADD_TEST(test_EVP_PKCS82PKEY);
 #endif
+    ADD_ALL_TESTS(test_set_get_raw_keys, OSSL_NELEM(keys) * 2);
```
Comment (richsalz) [inline test/evp_extra_test.c]: This seems a little needlessly tricky.  Perhaps add a wrapper function that calls the current test_set_get_raw_keys with an additional pub/private param, something like this:
```
ADD_ALL_TESTS(test_get_set, OSSL_NELEM(keys));
...
int test_get_set(int tst)
{
    return test_set_get_raw_keys(tst, 1) && test_set_get_raw_keys(tst, 0)
}
```
I think this makes the existing test routine smaller and a bit more clear, too.
```

File: crypto/asn1/ameth_lib.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/asn1/ameth_lib.c]: This is a function to help set up a new custom EVP_PKEY_ASN1_METHOD. Hence the ASN1 in the name (as is consistent with all the other functions above this). The name of the method itself is a misnomer. Originally this was about encoding and decoing to/from ASN.1. This has changed over time to encompass much more. Now its about constructing, checking, encoding, decoding and setting up keys and parameters ready for later use (which may include some ASN.1 work). It probably should have a new name, but that's how it is in our API so we're stuck with it for now.

**Comment by mattcaswell** [comment]
All of the above issues fixed, as well as fixing a couple of errors identified by travis. Fixup commits pushed.

**Comment by mattcaswell** [comment]
Pushed a couple more commits to address travis/appveyor failures.

**Comment by mattcaswell** [comment]
Grrr..still had some travis/appveyor failures. Now fixed (hopefully). @richsalz - please can you reconfirm?

**Comment by richsalz** [comment]
and also reconfirm.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**